src/Databases/PostgreSQL/DatabasePostgreSQLReplica.cpp

```diff
@@ -0,0 +1,234 @@
+#include <DataTypes/DataTypeArray.h>
+#include <Databases/DatabaseOrdinary.h>
+#include <Databases/DatabaseAtomic.h>
+#include <Storages/StoragePostgreSQL.h>
+#include <Interpreters/Context.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ParserCreateQuery.h>
+#include <Parsers/parseQuery.h>
+#include <Parsers/queryToString.h>
+#include <Common/escapeForFileName.h>
+#include <Poco/DirectoryIterator.h>
+#include <Poco/File.h>
+#include <Databases/PostgreSQL/fetchPostgreSQLTableStructure.h>
+#include <Common/Macros.h>
+#include <common/logger_useful.h>
+
+
+namespace DB
+{
+
+static const auto METADATA_SUFFIX = ".postgresql_replica_metadata";
+
+template<>
+DatabasePostgreSQLReplica<DatabaseOrdinary>::DatabasePostgreSQLReplica(
+        const Context & context,
+        const String & metadata_path_,
+        UUID /* uuid */,
+        const ASTStorage * database_engine_define_,
+        const String & database_name_,
+        const String & postgres_database_name,
+        PostgreSQLConnectionPtr connection_,
+        std::unique_ptr<PostgreSQLReplicaSettings> settings_)
+    : DatabaseOrdinary(
+            database_name_, metadata_path_, "data/" + escapeForFileName(database_name_) + "/",
+            "DatabasePostgreSQLReplica<Ordinary> (" + database_name_ + ")", context)
+    , log(&Poco::Logger::get("PostgreSQLReplicaDatabaseEngine"))
+    , global_context(context.getGlobalContext())
+    , metadata_path(metadata_path_)
+    , database_engine_define(database_engine_define_->clone())
+    , database_name(database_name_)
+    , remote_database_name(postgres_database_name)
+    , connection(std::move(connection_))
+    , settings(std::move(settings_))
+{
+}
+
+
+template<>
+DatabasePostgreSQLReplica<DatabaseAtomic>::DatabasePostgreSQLReplica(
+        const Context & context,
+        const String & metadata_path_,
+        UUID uuid,
+        const ASTStorage * database_engine_define_,
+        const String & database_name_,
+        const String & postgres_database_name,
+        PostgreSQLConnectionPtr connection_,
+        std::unique_ptr<PostgreSQLReplicaSettings> settings_)
+    : DatabaseAtomic(database_name_, metadata_path_, uuid, "DatabasePostgreSQLReplica<Atomic> (" + database_name_ + ")", context)
+    , global_context(context.getGlobalContext())
+    , metadata_path(metadata_path_)
+    , database_engine_define(database_engine_define_->clone())
+    , remote_database_name(postgres_database_name)
+    , connection(std::move(connection_))
+    , settings(std::move(settings_))
+{
+}
+
+
+template<typename Base>
+void DatabasePostgreSQLReplica<Base>::startSynchronization()
+{
+    replication_handler = std::make_unique<PostgreSQLReplicationHandler>(
+            remote_database_name,
+            connection->conn_str(),
+            metadata_path + METADATA_SUFFIX,
+            std::make_shared<Context>(global_context),
+            settings->postgresql_replica_max_block_size.changed
```

Comment:
I thought it would be better to use `max_insert_block_size` settings if postgresql replication specific max_block_size is not added. But I agree, it will be nicer to just add a proper default to `postgresql_replica_max_block_size`. Fixed.